### PR TITLE
Added bitbucket link to multibranch and workflow jobs

### PR DESCRIPTION
**Pull request by mhenschke-atl**
Addressing https://issues.jenkins-ci.org/browse/JENKINS-63070

In https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/154, we added a link in the freestyle job sidebar back to bitbucket, if the job had a bitbucket SCM configured. This extends that feature to pipeline jobs, and multibranch pipeline folders.

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import hudson.util.FormValidation;
+import jenkins.model.TransientActionFactory;
+import org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+
+@Extension
+public class BitbucketJobLinkActionFactory extends TransientActionFactory<Job> {
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java:17]: The key change here is the action factory is now opened for all _jobs_, not just projects. I'm not 100% with Jenkins standard, if they prefer more specific or generic action factories, and happy to produce two action factories for each job if that's preferred.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    private Optional<BitbucketSCM> getBitbucketSCM(Job job) {
+        if (job instanceof FreeStyleProject) {
+            FreeStyleProject freeStyleProject = (FreeStyleProject) job;
+            if (freeStyleProject.getScm() instanceof BitbucketSCM) {
+                return Optional.of((BitbucketSCM) freeStyleProject.getScm());
+            }
+        } else if (job instanceof WorkflowJob) {
+            WorkflowJob workflowJob = (WorkflowJob) job;
+            if (workflowJob.getDefinition() instanceof CpsScmFlowDefinition) {
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java:63]: This captures all Pipeline jobs being created, but does not include the branches of a Multibranch pipeline job. I did this because I didn't feel like you need the link in both places, the parent is usually enough, but it's a fairly trivial change if people disagree.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java [hunk 3]
```diff
@@ -3,24 +3,24 @@
...
-        if (!(target.getScm() instanceof BitbucketSCM)) {
+    public Collection<? extends Action> createFor(WorkflowMultiBranchProject target) {
+        Optional<BitbucketSCMSource> maybeSource = getBitbucketSCMSource(target);
+        if (!maybeSource.isPresent()) {
             return Collections.emptySet();
         }
-        BitbucketSCM bitbucketSCM = (BitbucketSCM) target.getScm();
-        BitbucketSCM.DescriptorImpl descriptor = (BitbucketSCM.DescriptorImpl) bitbucketSCM.getDescriptor();
+
+        BitbucketSCMSource bitbucketSCM = maybeSource.get();
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java:22]: This block of code is largely duplicated between each action factory because, unfortunately, there is no unifying interface between `BitbucketSCM`, `BitbucketSCMStep` and `BitbucketSCMSource` in the same way there is for our descriptors. If there was, we could put this in an inherited abstract class, but that would probably deserve it's own PR. Opinions?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java [hunk 4]
```diff
@@ -44,7 +44,14 @@
...
-        return Project.class;
+    public Class<WorkflowMultiBranchProject> type() {
+        return WorkflowMultiBranchProject.class;
+    }
+
+    private Optional<BitbucketSCMSource> getBitbucketSCMSource(WorkflowMultiBranchProject project) {
+        return project.getSCMSources().stream()
+                .filter(source -> source instanceof BitbucketSCMSource)
+                .map(source -> (BitbucketSCMSource) source)
+                .findAny();
```
Comment (EliSalter-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java:55]: is it possible for users to hook up multiple BbS instances to their Jenkins instance? I am not totally familiar with BitbucketSCMSource but as it has `mirrorName` are there any implications of this

Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java:55]: This is an interesting question. Re. mirrors, no they will always be associated with the same instance. _However_ customers can have multiple branch sources. We've never really tested this, I assume you would do this if for example you had a forked workflow across instances that need to share a build plan.

In this case, the link doesn't make a lot of sense, and it might be better if we don't provide one, because which "bitbucket" to go to is completely ambiguous. That's a strong argument for omitting the link from the multibranch job entirely and just having it available on pipeline branches instead.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java (hunk 3 above)
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java:22]: Re. my earlier comment I think I may end up completely removing this class instead, as the link location is ambiguous in forked workflows across instances. I agree a unifying interface will make such work easier in the future, but I think I prefer that to go to a separate PR- it's a sizeable refactor.

Comment (EliSalter-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java:22]: Fantastic and I 100% agree with placing that refactor in a seperate PR.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java (hunk 4 above)
Comment (EliSalter-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java:55]: Because we can have multiple branch sources it might be best to not provide a link if we can't break down that ambiguity. It would be interesting to see how the Branch Source Plugin (referenced in the ticket https://issues.jenkins-ci.org/browse/JENKINS-63070) navigates this and if they've come across similar issues.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        BitbucketSCM bitbucketSCM = maybeSCM.get();
+        BitbucketSCM.DescriptorImpl descriptor = (BitbucketSCM.DescriptorImpl) bitbucketSCM.getDescriptor();
+        Optional<BitbucketServerConfiguration> maybeConfig = descriptor.getConfiguration(bitbucketSCM.getServerId());
+        String serverId = Objects.toString(bitbucketSCM.getServerId(), "");
+        String credentialsId = Objects.toString(bitbucketSCM.getCredentialsId(), "");
+
+        FormValidation configValid = FormValidation.aggregate(Arrays.asList(
+                maybeConfig.map(BitbucketServerConfiguration::validate).orElse(FormValidation.error("Config not present")),
+                descriptor.doCheckProjectName(serverId, credentialsId, bitbucketSCM.getProjectName()),
+                descriptor.doCheckRepositoryName(serverId, credentialsId, bitbucketSCM.getProjectName(), bitbucketSCM.getRepositoryName())
```
[the remark this answers is not in the document]
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java:35]: Hmm I think I personally prefer to keep this concise (given the call is free, and the project name is only used on those two lines)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java (hunk 3 above)
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java]: So good news- by injecting the form validator and plugin configuration, and using the bitbucket repository instead, we eschew the need for such a refactor- all the information we require is contained therein, and we should use it instead of making new interfaces.

**Comment by mhenschke-atl** [comment]
So the changed goal of this PR is to have Bitbucket links appear on all _jobs_ that we support (not folders or multibranch containers). There are a few cases we consider:
- Freestyle jobs we extract the link from the BitbucketSCM directly
- Pipeline jobs we extract from the SCM as well
- Multibranch pipeline jobs built using an SCM Step will store the SCM that was used to build them, if they have at least 1 completed build against them. There can be multiple (your build might checkout several repos as part of a pipeline). I'm directing them to the first checkout Jenkins knows about. We need this case in the event the repository and jenkinsfile are stored on separate instances (though this is very much an edge case)
- Multibranch pipeline jobs with no build against it, or built with a `checkout scm` step has to consult it's parents' BitbucketSCM source. _We do not support multiple SCM sources_, so this plugin takes the first and assumes it is correct.

**Comment by mhenschke-atl** [comment]
This is not a perfect solution- it does not support the case where a build pipeline uses multiple SCM steps to perform checkouts in Bitbucket, and the first one is not the repo of code that is being built. I can think of cases where this might happen- you checkout and prepare a testing suite as part of a pipeline, then checkout the code itself and run the suite against it. It's a small case though, so at the risk of making the solution more complex I think this is okay, and we can address issues as they arise in the issue tracker.

**Review by elisalter — APPROVED** [review]
Looks good! 👍

**Review by dkjellin — APPROVED** [review]

**Review by EliSalter-atl — APPROVED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        if (configValid.kind == FormValidation.Kind.ERROR) {
+            return Collections.emptySet();
+        }
+
+        String url = maybeConfig.get().getBaseUrl() +
+                     "/projects/" +
+                     bitbucketRepository.getProjectKey() +
+                     "/repos/" +
+                     bitbucketRepository.getRepositorySlug();
+        return Collections.singleton(BitbucketExternalLink.createDashboardLink(url, target));
```
Comment (KalleOlaviNiemitalo) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java:66]: Does this make all links point to the repository in Bitbucket Server? In [](https://issues.jenkins-ci.org/browse/JENKINS-63070), I hoped that each Jenkins job corresponding to a branch would link to that branch in Bitbucket Server, and likewise with pull requests. Is that implemented in some other method?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketExternalLink.java [hunk 7]
```diff
@@ -50,7 +49,7 @@ public String getIconClassName() {
     @CheckForNull
     @Override
     public String getIconFileName() {
-        if (!project.hasPermission(Item.CONFIGURE)) {
+        if (!owner.hasPermission(Item.CONFIGURE)) {
             return null;
         }
```
Comment (KalleOlaviNiemitalo) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketExternalLink.java:54]: I don't see the point of this CONFIGURE permission check. It was not added in this PR though.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java (hunk 6 above)
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java:66]: Thanks for the comments Kalle, I will address these in a separate PR. I didn't include the main workflow project because it's possible to have multiple branch sources (and I didn't want to flood the project page with links)- chatting with team mates, it seems like this is possible, but not a solution we support so I'll make the changes as requested.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketExternalLink.java (hunk 7 above)
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketExternalLink.java:54]: This was awhile back but I believe the thinking here was the server the job belongs to was part of the config. That said I agree it's not necessary. Will remove in a separate PR.
